### PR TITLE
[clang-tidy] fix bugprone-sizeof-expression when sizeof expression with template types

### LINKED ISSUE #115175: [clang-tidy] bugprone-sizeof-expression false positives
```c++
auto foo(auto t) {
  return sizeof(t.val) / sizeof(t.val[0]);
}
```

https://godbolt.org/z/4qa1odf5W

> ```c++
> <source>:2:24: warning: suspicious usage of 'sizeof(...)/sizeof(...)'; both expressions have the same type [bugprone-sizeof-expression]
>     2 |   return sizeof(t.val) / sizeof(t.val[0]);
>       |          ~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~
> 1 warning generated.
> ```

> clang version 20.0.0git (https://github.com/llvm/llvm-project.git 4d374479bea4b33c5623ccfedc0870e396fc34cd)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/compiler-explorer/clang-trunk-20241106/bin

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixed: #115175.
`dependent type` are not the same even pointers are the same.

### CODE REVIEW

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -175,7 +175,8 @@ Changes in existing checks
 - Improved :doc:`bugprone-sizeof-expression
   <clang-tidy/checks/bugprone/sizeof-expression>` check to find suspicious
   usages of ``sizeof()``, ``alignof()``, and ``offsetof()`` when adding or
-  subtracting from a pointer directly or when used to scale a numeric value.
+  subtracting from a pointer directly or when used to scale a numeric value and
+  fix false positive when sizeof expression with template types.
```
Comment (whisperity) [inline clang-tools-extra/docs/ReleaseNotes.rst:179]: If there were no releases since this change, then there is no need for the release note entry. Basically you're hotfixing a change that is coming out in the same hypothetical next release.

If the change in this patch is not related to what is mentioned in this changelog entry, then it is worth a full new changelog entry, like this:

```suggestion
  subtracting from a pointer directly or when used to scale a numeric value.
  
- Fixed :doc:`bugprone-sizeof-expression
  <clang-tidy/checks/bugprone/sizeof-expression>` check not to report on
  ``sizeof()/sizeof()`` expressions for template types.
```

File: clang-tools-extra/clang-tidy/bugprone/SizeofExpressionCheck.cpp [hunk 2]
```diff
@@ -400,7 +400,9 @@ void SizeofExpressionCheck::check(const MatchFinder::MatchResult &Result) {
            "suspicious usage of 'sizeof(array)/sizeof(...)';"
            " denominator differs from the size of array elements")
           << E->getLHS()->getSourceRange() << E->getRHS()->getSourceRange();
-    } else if (NumTy && DenomTy && NumTy == DenomTy) {
+    } else if (NumTy && DenomTy && NumTy == DenomTy &&
+               !NumTy->isDependentType()) {
+      // dependent type should not be compared.
```
Comment (whisperity) [inline clang-tools-extra/clang-tidy/bugprone/SizeofExpressionCheck.cpp:405]:
```suggestion
      // Dependent type should not be compared.
```

(Although not sure what informational value this comment has, it might as well could be removed completely.)

**Review by whisperity — CHANGES_REQUESTED** [review]

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 1 above)
Comment (HerrCai0907) [reply clang-tools-extra/docs/ReleaseNotes.rst:179]: I think not. clang-tidy release-note always combines the all changes of rule in one item instead of spilt them.

Comment (HerrCai0907) [reply clang-tools-extra/docs/ReleaseNotes.rst:179]: example: 
`readability-container-contains`
`readability-enum-initial-value`

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/sizeof-expression.cpp [hunk 3]
```diff
@@ -385,3 +385,8 @@ int ValidExpressions() {
   sum += sizeof(PtrArray) / sizeof(A[0]);
   return sum;
 }
+
+template<class T>
+int ValidateTemplateTypeExpressions(T t) {
+  return sizeof(t.val) / sizeof(t.val[0]);
+}
```
Comment (5chmidti) [inline clang-tools-extra/test/clang-tidy/checkers/bugprone/sizeof-expression.cpp:392]: Please note the issue somewhere on this test-case (115175) (e.g. namespace)

**Review by 5chmidti — APPROVED** [review]

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 1 above)
Comment (5chmidti) [reply clang-tools-extra/docs/ReleaseNotes.rst:179]: > If there were no releases since this change, then there is no need for the release note entry. Basically you're hotfixing a change that is coming out in the same hypothetical next release.

I'm not sure what you meant here. There should be a release note, unless the issue was introduced in the current, release cycle. The diagnostic reproduces in clang-tidy 18 for me, or did you mean something different (e.g., general *in-case-of* comment)?

> clang-tidy release-note always combines the all changes of rule in one item instead of spilt them.

+1, that's the current approach

Comment (whisperity) [reply clang-tools-extra/docs/ReleaseNotes.rst:179]: > > If there were no releases since this change, then there is no need for the release note entry. Basically you're hotfixing a change that is coming out in the same hypothetical next release.
> 
> I'm not sure what you meant here. There should be a release note, unless the issue was introduced in the current, release cycle. The diagnostic reproduces in clang-tidy 18 for me, or did you mean something different (e.g., general _in-case-of_ comment)?

Because of the merged release note I thought this is related to #106061, hence the confusion.
